File: src/diagnosis/classify.ts
// Deterministic classification per flow. Each flow maps its raw signal/error
// code onto ONE narrow reason bucket. Pure functions, no LLM.

import { FlowEvent, FlowType, ReasonBucket } from '../flows/types';

// Enum of our finite reason buckets per flow (narrow taxonomy)
export const PAYMENT_BUCKETS: readonly ReasonBucket[] = [
  'insufficient_funds',
  'card_expired',
  'mandate_revoked',
  'bank_declined_transient',
  'auth_3ds_abandoned',
  'fraud_flagged',
];

// ---------------------------------------------------------------------------
// Razorpay error-code map (shared by subscription + mandate + payment flows)
// ---------------------------------------------------------------------------
const RAZORPAY_ERROR_CODES: Record<string, ReasonBucket> = {
  BAD_REQUEST_ERROR: 'insufficient_funds',
  CARD_DECLINED: 'insufficient_funds',
  CARD_EXPIRED: 'card_expired',
  CARD_EXPIRED_CODE: 'card_expired',
  DIRECTORAUTHPARAM_FAILED: 'auth_3ds_abandoned',
  ISSUER_UNAVAILABLE: 'bank_declined_transient',
  UNAUTHORIZED_PAYMENT: 'bank_declined_transient',
  PAYMENT_AUTHENTICATION_FAILED: 'bank_declined_transient',
  BANK_REFUND_DECLINED: 'bank_declined_transient',
  NETWORK_ERROR: 'bank_declined_transient',
  MANDATE_REVOKED: 'mandate_revoked',
  MANDATE_INVALID: 'mandate_revoked',
  MANDATE_CANCELLED: 'mandate_revoked',
  FRAUD_DETECTED: 'fraud_flagged',
  RISK_DECLINE: 'fraud_flagged',
  AUTH_FAILED: 'auth_3ds_abandoned',
  CUSTOMER_ABANDONED: 'auth_3ds_abandoned',
  AUTH_ACCEPTED_LATER: 'auth_3ds_abandoned',
  SUCCESS_RATE_DROP: 'success_rate_drop',
  LATENCY_SPIKE: 'latency_spike',
  ISSUER_DOWN: 'issuer_down',
};

interface Signal {
  [k: string]: unknown;
}

function pickBucketFromCode(code: string | undefined): ReasonBucket | undefined {
  if (!code) return undefined;
  return RAZORPAY_ERROR_CODES[code.toUpperCase()];
}

// ---------------------------------------------------------------------------
// Flow 3: failed subscription / mandate retry
// ---------------------------------------------------------------------------
export function classifySubscription(signal: Signal): ReasonBucket {
  const code = pickBucketFromCode(signal.error_code as string);
  if (code) return code;
  const desc = `${signal.error_description ?? ''} ${signal.error_reason ?? ''}`.toLowerCase();
  if (desc.includes('insufficient') || desc.includes('limit')) return 'insufficient_funds';
  if (desc.includes('expired')) return 'card_expired';
  if (desc.includes('mandate') && (desc.includes('revoke') || desc.includes('cancel') || desc.includes('invalid')))
    return 'mandate_revoked';
  if (desc.includes('fraud') || desc.includes('risk') || desc.includes('suspicious')) return 'fraud_flagged';
  if (desc.includes('abandoned') || desc.includes('3ds') || desc.includes('auth')) return 'auth_3ds_abandoned';
  return 'bank_declined_transient';
}

// ---------------------------------------------------------------------------
// Flow 2: checkout abandonment
// ---------------------------------------------------------------------------
export function classifyCheckout(signal: Signal): ReasonBucket {
  const step = (signal.abandoned_at_step as string) ?? '';
  const err = (signal.error as string) ?? '';
  if (err && err.length > 0) return 'checkout_error';
  if (step.includes('payment') || step.includes('otp') || step.includes('upi')) return 'payment_step_abandoned';
  if (step.includes('address')) return 'address_abandoned';
  if ((signal.page_load_ms as number) > 5000) return 'slow_checkout';
  if (signal.price_mismatch === true) return 'price_mismatch';
  return 'payment_step_abandoned';
}

// ---------------------------------------------------------------------------
// Flow 4: B2B receivables
// ---------------------------------------------------------------------------
export function classifyReceivable(signal: Signal): ReasonBucket {
  const disputed = signal.disputed === true || (signal.dispute_note && (signal.dispute_note as string).length > 0);
  if (disputed) return 'disputed_receivable';
  const overdueDays = (signal.overdue_days as number) ?? 0;
  if (overdueDays < 60) return 'overdue_net30';
  return 'overdue_net60';
}

// ---------------------------------------------------------------------------
// Flow 1: payment degradation
// ---------------------------------------------------------------------------
export function classifyDegradation(signal: Signal): ReasonBucket {
  const code = pickBucketFromCode(signal.error_code as string);
  if (code) return code;
  if ((signal.success_rate as number) < 0.85) return 'success_rate_drop';
  if ((signal.latency_ms as number) > 2000) return 'latency_spike';
  if (signal.issuer_down === true) return 'issuer_down';
  return 'bank_declined_transient';
}

// ---------------------------------------------------------------------------
// Flow 6: Hinglish voice
// ---------------------------------------------------------------------------
export function classifyVoice(signal: Signal): ReasonBucket {
  const state = (signal.call_state as string) ?? '';
  if (state === 'missed' || state === 'no_answer') return 'voice_missed_call';
  if (state === 'call_back_requested') return 'voice_asked_call_back';
  return 'voice_unreachable';
}

// ---------------------------------------------------------------------------
// Flow 7: Promise-to-pay tracker
// ---------------------------------------------------------------------------
export function classifyPromiseToPay(signal: Signal): ReasonBucket {
  const status = (signal.ptp_status as string) ?? '';
  if (status === 'broken') return 'ptp_broken';
  if (status === 'missed') return 'ptp_missed';
  if (status === 'committed') return 'ptp_committed';
  return 'ptp_committed';
}

// Dispatch: normalize any FlowEvent's signal into one reason bucket.
export function classify(event: FlowEvent): ReasonBucket {
  const signal = event.signal ?? {};
  switch (event.flow) {
    case 'failed_subscription':
    case 'mandate_retry':
      return classifySubscription(signal);
    case 'checkout_abandonment':
      return classifyCheckout(signal);
    case 'b2b_receivables':
      return classifyReceivable(signal);
    case 'payment_degradation':
      return classifyDegradation(signal);
    case 'hinglish_voice':
      return classifyVoice(signal);
    case 'promise_to_pay':
      return classifyPromiseToPay(signal);
  }
}

File: go/rz/classify.go
package rz

import (
	"strings"
)

// Deterministic classification per flow. Each flow maps its raw signal/error
// code onto ONE narrow reason bucket. Pure functions, no LLM.

var paymentBuckets = []ReasonBucket{
	ReasonInsufficientFunds,
	ReasonCardExpired,
	ReasonMandateRevoked,
	ReasonBankDeclinedTransien,
	ReasonAuth3dsAbandoned,
	ReasonFraudFlagged,
}

// Razorpay error-code map (shared by subscription + mandate + payment flows).
var razorpayErrorCodes = map[string]ReasonBucket{
	"BAD_REQUEST_ERROR":             ReasonInsufficientFunds,
	"CARD_DECLINED":                 ReasonInsufficientFunds,
	"CARD_EXPIRED":                  ReasonCardExpired,
	"CARD_EXPIRED_CODE":             ReasonCardExpired,
	"DIRECTORAUTHPARAM_FAILED":      ReasonAuth3dsAbandoned,
	"ISSUER_UNAVAILABLE":            ReasonBankDeclinedTransien,
	"UNAUTHORIZED_PAYMENT":          ReasonBankDeclinedTransien,
	"PAYMENT_AUTHENTICATION_FAILED": ReasonBankDeclinedTransien,
	"BANK_REFUND_DECLINED":          ReasonBankDeclinedTransien,
	"NETWORK_ERROR":                 ReasonBankDeclinedTransien,
	"MANDATE_REVOKED":               ReasonMandateRevoked,
	"MANDATE_INVALID":               ReasonMandateRevoked,
	"MANDATE_CANCELLED":             ReasonMandateRevoked,
	"FRAUD_DETECTED":                ReasonFraudFlagged,
	"RISK_DECLINE":                  ReasonFraudFlagged,
	"AUTH_FAILED":                   ReasonAuth3dsAbandoned,
	"DIRECTAUTHPARAM_FAILED":        ReasonAuth3dsAbandoned, // canonical Razorpay code
	"CUSTOMER_ABANDONED":            ReasonAuth3dsAbandoned,
	"AUTH_ACCEPTED_LATER":           ReasonAuth3dsAbandoned,
	"SUCCESS_RATE_DROP":             ReasonSuccessRateDrop,
	"LATENCY_SPIKE":                 ReasonLatencySpike,
	"ISSUER_DOWN":                   ReasonIssuerDown,
}

// strField reads a string field from a signal map.
func strField(s map[string]any, key string) string {
	v, ok := s[key]
	if !ok || v == nil {
		return ""
	}
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}

func numField(s map[string]any, key string) float64 {
	v, ok := s[key]
	if !ok || v == nil {
		return 0
	}
	switch n := v.(type) {
	case float64:
		return n
	case float32:
		return float64(n)
	case int:
		return float64(n)
	case int64:
		return float64(n)
	case int32:
		return float64(n)
	case uint64:
		return float64(n)
	}
	return 0
}

func boolField(s map[string]any, key string) bool {
	v, ok := s[key]
	if !ok || v == nil {
		return false
	}
	if b, ok := v.(bool); ok {
		return b
	}
	return false
}

func pickBucketFromCode(code string) (ReasonBucket, bool) {
	if code == "" {
		return "", false
	}
	b, ok := razorpayErrorCodes[strings.ToUpper(code)]
	return b, ok
}

// ClassifySubscription is Flow 3: failed subscription / mandate retry.
func ClassifySubscription(signal map[string]any) ReasonBucket {
	if code, ok := pickBucketFromCode(strField(signal, "error_code")); ok {
		return code
	}
	desc := strings.ToLower(strField(signal, "error_description") + " " + strField(signal, "error_reason"))
	switch {
	case strings.Contains(desc, "insufficient") || strings.Contains(desc, "limit"):
		return ReasonInsufficientFunds
	case strings.Contains(desc, "expired"):
		return ReasonCardExpired
	case strings.Contains(desc, "mandate") && (strings.Contains(desc, "revoke") || strings.Contains(desc, "cancel") || strings.Contains(desc, "invalid")):
		return ReasonMandateRevoked
	case strings.Contains(desc, "fraud") || strings.Contains(desc, "risk") || strings.Contains(desc, "suspicious"):
		return ReasonFraudFlagged
	case strings.Contains(desc, "abandoned") || strings.Contains(desc, "3ds") || strings.Contains(desc, "auth"):
		return ReasonAuth3dsAbandoned
	default:
		return ReasonBankDeclinedTransien
	}
}

// ClassifyCheckout is Flow 2: checkout abandonment.
func ClassifyCheckout(signal map[string]any) ReasonBucket {
	step := strField(signal, "abandoned_at_step")
	err := strField(signal, "error")
	if err != "" {
		return ReasonCheckoutError
	}
	if strings.Contains(step, "payment") || strings.Contains(step, "otp") || strings.Contains(step, "upi") {
		return ReasonPaymentStepAbandoned
	}
	if strings.Contains(step, "address") {
		return ReasonAddressAbandoned
	}
	if numField(signal, "page_load_ms") > 5000 {
		return ReasonSlowCheckout
	}
	if boolField(signal, "price_mismatch") {
		return ReasonPriceMismatch
	}
	return ReasonPaymentStepAbandoned
}

// ClassifyReceivable is Flow 4: B2B receivables.
func ClassifyReceivable(signal map[string]any) ReasonBucket {
	if boolField(signal, "disputed") || strField(signal, "dispute_note") != "" {
		return ReasonDisputedReceivable
	}
	overdueDays := int(numField(signal, "overdue_days"))
	if overdueDays < 60 {
		return ReasonOverdueNet30
	}
	return ReasonOverdueNet60
}

// ClassifyDegradation is Flow 1: payment degradation.
func ClassifyDegradation(signal map[string]any) ReasonBucket {
	if code, ok := pickBucketFromCode(strField(signal, "error_code")); ok {
		return code
	}
	if numField(signal, "success_rate") < 0.85 {
		return ReasonSuccessRateDrop
	}
	if numField(signal, "latency_ms") > 2000 {
		return ReasonLatencySpike
	}
	if boolField(signal, "issuer_down") {
		return ReasonIssuerDown
	}
	return ReasonBankDeclinedTransien
}

// ClassifyVoice is Flow 6: Hinglish voice.
func ClassifyVoice(signal map[string]any) ReasonBucket {
	state := strField(signal, "call_state")
	switch state {
	case "missed", "no_answer":
		return ReasonVoiceMissedCall
	case "call_back_requested":
		return ReasonVoiceAskedCallBack
	default:
		return ReasonVoiceUnreachable
	}
}

// ClassifyPromiseToPay is Flow 7: promise-to-pay tracker.
func ClassifyPromiseToPay(signal map[string]any) ReasonBucket {
	status := strField(signal, "ptp_status")
	switch status {
	case "broken":
		return ReasonPtpBroken
	case "missed":
		return ReasonPtpMissed
	default:
		return ReasonPtpCommitted
	}
}

// Classify normalizes any FlowEvent's signal into one reason bucket.
func Classify(event *FlowEvent) ReasonBucket {
	signal := event.Signal
	if signal == nil {
		signal = map[string]any{}
	}
	switch event.Flow {
	case FlowFailedSubscription, FlowMandateRetry:
		return ClassifySubscription(signal)
	case FlowCheckoutAbandonment:
		return ClassifyCheckout(signal)
	case FlowB2BReceivables:
		return ClassifyReceivable(signal)
	case FlowPaymentDegradation:
		return ClassifyDegradation(signal)
	case FlowHinglishVoice:
		return ClassifyVoice(signal)
	case FlowPromiseToPay:
		return ClassifyPromiseToPay(signal)
	}
	return ReasonBankDeclinedTransien
}
